### ISSUE TITLE
Unable to fetch Firewall Rules via firewall-plugin API

### ISSUE THREAD

**Issue by peironem** [opnsense/plugins#1898]
Hi.

I'm having some troubles trying to fetch fw rules.
I read the dock and looked at the examples at the page https://docs.opnsense.org/development/api/plugins/firewall.html but i still have problems.
I'm trying to use the searchRule API via python, and i can't get anything back with a simple call without any attributes, here's my code

```python
import json, requests


KEY = "my_key"
SECRET = "my_secret"

url = "https://192.168.2.1/api/firewall/filter/searchRule"
r = requests.get(url, verify=False, auth=(KEY, SECRET))
print(r.text)
```

result is always 
```python
{"rows":[],"rowCount":0,"total":0,"current":1}
```

Well, i've tryed to add some attributes to the call, but nothing changes, notice that i setted "icmp_dummies_rule" as the rule description cause i thought taht searchPhrase matches that field (also there's no unique name filed available, as far as i can see)
I tryed:
```python
url = 'https://192.168.2.1/api/firewall/filter/searchRule?searchPhrase="icmp_dummies_rule"'
url = 'https://192.168.2.1/api/firewall/filter/searchRule?searchPhrase=icmp_dummies_rule'
url = 'https://192.168.2.1/api/firewall/filter/searchRule?current=1&searchPhrase="icmp_dummies_rule"'
url = 'https://192.168.2.1/api/firewall/filter/searchRule?current=1&searchPhrase=icmp_dummies_rule'
```

Also i can't pass any UUID cause i don't know the rule UUID (but i prefere to retrive all the rules for flexibility reason and cleaner code).
Any idea here?
Thanks.

**Comment by peironem** [comment]
I'm sorry.
What "support" label stands for?
Have i wrote in the wrong place?

**Comment by AdSchellevis** [comment]
It stands for community support. You can ask questions on GitHub (although we prefer our forum at https://forum.opnsense.org for questions), we just label them accordingly.

**Comment by wrobelda** [comment]
This shouldn't be closed, I have same exact issue here. The examples simply don't work at all.

**Comment by AdSchellevis** [comment]
@wrobelda you're free to open a PR to discuss improvements at any time, so far your contributions haven't been very helpful to be honest.

**Comment by wrobelda** [comment]
@AdSchellevis well, this is an issue tracker, you really expect every user to only file the PRs when they have a _meaningful_ issue to report? Do you really expect all of OPNSense users to be software engineers? 

@peironem  has spent their own time creating a meaningful report and this was labeled as "support" without a word from the maintainers, without any testing, having been sent off to the Forums, where issues like these are regularly left without any response whatsoever, e.g.: https://forum.opnsense.org/index.php?topic=30810.msg150184#msg150184

So what I am simply asking is what are the criteria that decide which of the issues are labeled as "support" or "bug"?

**Comment by wrobelda** [comment]
> @wrobelda as an API engineer it probably helps a bit to know something about software engineering. Our policies are clearly described in our contribution guidelines (https://github.com/opnsense/plugins/blob/master/CONTRIBUTING.md). For small projects with less interaction it might work to validate everything, in reality for large scale projects one has to prioritise and traction depends on user feedback and likeliness of something being a bug.

That's understandable. But even knowing a bit, or even more than a bit, about software engineering, won't help if the documentation is lacking or straight up misleading, like in the case of firewall APIs docs. 

There were a bunch of issues reported here or in the forums around the API usage and brushing them all off by assuming "lack of knowledge of the requestor" comes off as patronizing, to say the least. I personally would instead assume that something is maybe off around the documentation itself, which it clearly was, as even with having 15 years of experience in engineering I had a super hard time understanding not *how* to use it, but how to use it with *expected results*. 

Hopefully the pull requests I filed around the docs make a strong argument that the docs themselves were, indeed, not clear enough around the peculiarities of the Firewall API.

> This one by the way is quite easy to test from your browser (https://xxx.xxx.xxx.xxx/api/firewall/filter/searchRule?current=1&rowCount=7), which looks good on my end. 

OK. So it works for you, but frankly it's still surprising that you find the "works for me" a good enough response, and that you didn't consider that there might be some deeper issue involved:  like the fact that all of the people having trouble with Firewall API perhaps didn't grasp that *it doesn't work with pre-existing rules*. Something so fundamental yet completely absent from the documentation (again, as exemplified by opnsense/docs#437 PR).

> Please do enjoy the rest of your day and try to understand our position.

I do understand it, you're a small team working on a big project, but frankly you're not the only OSS project of that nature, but the maybe a tad too preemptive label-stamping may, in the long run, discourage the community from spending their own time on meticulously reporting issue and just move on (or back) to the "other" project.

**Comment by AdSchellevis** [comment]
> That's understandable. But even knowing a bit, or even more than a bit, about software engineering, won't help if the documentation is lacking or even misleading, like in the case of firewall APIs docs.

Let's agree everywhere there are assumptions, on your end, on ours, if the steps in a ticket are very clear (what did you do and did you expect) chances are more likely someone will point you in the right direction. At the end of the day our time is limited, labeling and automation is the only way to balance the work realistically and leveling expectations. Most people don't seem to understand the amount of time we as a team spend on this product, when it comes to plugins I would argue that the ones we did offer ourselves are more elaborately documented than the ones we didn't (e.g. https://docs.opnsense.org/development/api/plugins/firewall.html#concept in this case).

> Hopefully the pull requests I filed around the docs make a strong argument that the docs themselves were, indeed, not clear enough around the peculiarities of the Firewall API.

I don't mind making things more explicit, as mentioned earlier in the PR as well. If it helps people it's ok, but in reality it will never be totally clear as it lies in the eye of the beholder and the background he or she has.

> OK. So it works for you, but frankly it's still surprising that you find the "works for me" a good enough response, and that you didn't consider that there might be some deeper issue involved: like the fact that all of the people having trouble with Firewall API perhaps didn't grasp that it doesn't work with pre-existing rules. Something so fundamental yet completely absent from the documentation (again, as exemplified by https://github.com/opnsense/docs/pull/437 PR).

In fairness, this is  strange statement compared to your earlier feedback.
> .... I am an OSS maintainer myself and would never just close an issue without making sure that things "work for me".

I also don't agree "all of the people having trouble"  is statement reflecting reality, we know quite some people are using the api offered in this particular case.

> I do understand it, you're a small team working on a big project, but frankly you're not the only OSS project of that nature, and the — maybe a tad too preemptive — label-stamping may, in the long run, discourage the community from spending their own time on meticulously reporting issues and instead just move on (or back) to the "other" project.

I don't think we're doing too bad, but that too is in the eye of the beholder.

> I appreciate your input. Have a great day forward as well.

Likewise

**Comment by fichtner** [comment]
Probably not healthy, but I want to reference the following:

> I do understand it, you're a small team working on a big project, but frankly you're not the only OSS project of that nature, and the — maybe a tad too preemptive — label-stamping may, in the long run, **discourage the community from spending their own time on meticulously reporting issues and instead just move on (or back) to the "other" project.**

It might appear so to you, but I think you are basing this on a number of assumptions:

1. Comparable APIs exist in all projects.
2. Documentation (about 1 in particular?) is complete elsewhere and here it is not.
3. The community culture forces people to leave for a different project. As a broad statement it might contradict what you just implied for the case of 1 and 2.
4. BSD licensing exist for a reason and we, unlike others, have never strayed from this principle. It has upsides and downsides for authors as well as users.
5. We try to keep things technical and avoid unworkable solutions or unrealistic expectations. Keeping it simple and effective is a key strategy and it works for the project and the community alike.


Cheers,
Franco

**Comment by theAWSguy** [comment]
For anyone who hits this in 2025, I had everything working up to a while ago; I noticed the GUI change under Firewall > Automation and couldn't find my rules anymore.  I realized I needed to filter by interface.  Using this thread i was able to figure out I needed to add "?interface=opt7" to the end of the URL to be something like:  "{BASE_URL}/firewall/filter/searchRule/?interface=opt7".

Now when I test using my browser everything looks good and when I view my response.json() I see what my browser shows.
